src/components/sidenav/sidenav.ts

```diff
@@ -0,0 +1,353 @@
+      return this.openPromise_;
+    } else {
+      if (!this.closePromise_) {
+        this.closePromise_ = new Promise<void>((resolve, reject) => {
+          this.closePromiseResolve_ = resolve;
+          this.closePromiseReject_ = reject;
+        });
+      }
+      return this.closePromise_;
+    }
+  }
+
+
+  /**
+   * When transition has finished, set the internal state for classes and emit the proper event.
+   * The event passed is actually of type TransitionEvent, but that type is not available in
+   * Android so we use any.
+   * @param e The event.
+   * @private
+   */
+  @HostListener('transitionend', ['$event']) private onTransitionEnd_(e: Event) {
+    if (e.target == this.elementRef_.nativeElement
+        // Simpler version to check for prefixes.
+        && (e.propertyName.endsWith('transform')) {
+      this.transition_ = false;
+      if (this.opened_) {
+        if (this.openPromise_) {
+          this.openPromiseResolve_();
+        }
+        if (this.closePromise_) {
+          this.closePromiseReject_();
+        }
+
+        this.onOpen.emit(null);
+      } else {
+        if (this.closePromise_) {
+          this.closePromiseResolve_();
+        }
+        if (this.openPromise_) {
+          this.openPromiseReject_();
+        }
+
+        this.onClose.emit(null);
+      }
+
+      this.openPromise_ = null;
+      this.closePromise_ = null;
+    }
+  }
+
+  @HostBinding('class.md-sidenav-closing') private get isClosing_() {
+    return !this.opened_ && this.transition_;
+  }
+  @HostBinding('class.md-sidenav-opening') private get isOpening_() {
+    return this.opened_ && this.transition_;
+  }
+  @HostBinding('class.md-sidenav-closed') private get isClosed_() {
+    return !this.opened_ && !this.transition_;
+  }
+  @HostBinding('class.md-sidenav-opened') private get isOpened_() {
+    return this.opened_ && !this.transition_;
+  }
+  @HostBinding('class.md-sidenav-end') private get isEnd_() {
+    return this.align == 'end';
+  }
+  @HostBinding('class.md-sidenav-side') private get modeSide_() {
+    return this.mode == 'side';
+  }
+  @HostBinding('class.md-sidenav-over') private get modeOver_() {
+    return this.mode == 'over';
+  }
+  @HostBinding('class.md-sidenav-push') private get modePush_() {
+    return this.mode == 'push';
+  }
+
+  /**
+   * This is public because we need it from MdSidenavLayout, but it's undocumented and should
+   * not be used outside.
+   * @returns {number}
+   * @private
+   */
+  public get width_() {
+    if (this.elementRef_.nativeElement) {
+      return this.elementRef_.nativeElement.offsetWidth;
+    }
+    return 0;
+  }
+
+  private transition_: boolean = false;
+  private openPromise_: Promise<void>;
+  private openPromiseResolve_: () => void;
+  private openPromiseReject_: () => void;
+  private closePromise_: Promise<void>;
+  private closePromiseResolve_: () => void;
+  private closePromiseReject_: () => void;
+}
+
+
+
+/**
+ * <md-sidenav-layout> component.
+ */
```

Comment:
Maybe add a comment like "Container component that lays out the sidenav along with the main content."